File: src/me/shkschneider/skeleton/Hash.java
package me.shkschneider.skeleton;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

@SuppressWarnings("unused")
public class Hash {

    public static final String MD5 = "MD5";
    public static final String SHA = "SHA";
    public static final Integer MD5_LENGTH = 32;
    public static final Integer SHA_LENGTH = 40;

    protected static String hash(final String algorithm, final String string, final Integer length) {
        try {
            final MessageDigest messageDigest = MessageDigest.getInstance(algorithm);
            messageDigest.reset();
            messageDigest.update(string.getBytes());

            final StringBuilder stringBuilder = new StringBuilder();
            stringBuilder.setLength(0);

            final byte digest[] = messageDigest.digest();
            for (final byte d : digest) {
                final int b = d & 255;
                if (b < length) {
                    stringBuilder.append('0');
                }
                stringBuilder.append(Integer.toHexString(b));
            }
            return stringBuilder.toString();
        }
        catch (NoSuchAlgorithmException e) {
            Log.e("NoSuchAlgorithmException: " + e.getMessage());
        }
        return null;
    }

    public static String md5(final String string) {
        return hash(MD5, string, MD5_LENGTH);
    }

    public static String sha(final String string) {
        return hash(SHA, string, SHA_LENGTH);
    }

}

File: kotlin/src/main/kotlin/me/shkschneider/skeleton/kotlin/datax/Hash.kt
package me.shkschneider.skeleton.kotlin.datax

import me.shkschneider.skeleton.kotlin.log.Logger
import java.math.BigInteger
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

object Hash {

    private const val MD5 = "MD5"
    private const val SHA1 = "SHA-1"
    private const val SHA256 = "SHA-256"
    private const val SHA512 = "SHA-512"

    private fun hash(algorithm: String, string: String): String? =
        try {
            val messageDigest = MessageDigest.getInstance(algorithm).digest(string.toByteArray())
            BigInteger(1, messageDigest).toString(16).padStart(32, '0')
        } catch (e: NoSuchAlgorithmException) {
            Logger.wtf(e)
            null
        }

    @Deprecated("Unsafe.")
    fun md5(string: String): String? =
        hash(MD5, string)

    @Deprecated("Unsafe.")
    fun sha1(string: String): String? =
        hash(SHA1, string)

    fun sha256(string: String): String? =
        hash(SHA256, string)

    fun sha512(string: String): String? =
        hash(SHA512, string)

}
